### PR TITLE
[ENH] utils for programmatic soft dependency isolation

### LINKED ISSUE #7701: [ENH] utils for programmtic soft dependency isolation of `torch` and `transformers`
**Is your feature request related to a problem? Please describe.**

Yes. During the vendoring of a library in `sktime`, we need to often use soft dependencies like `torch` and `transformers` and we can't import them directly so we need to find a way around this. In the past, dummy classes were made at the file level in the vendored library but there is a downside to this method, every time a new library is vendored, the dummy classes are created again and again.

**Describe the solution you'd like**

Having utilities for these soft dependency and classes in the `util` directory will make it easier to import them from there and it will reduce confusion for developers.

### PR DESCRIPTION
**Pull request by jgyasu**
#### Reference Issues/PRs

Fixes #7701 

#### What does this implement/fix? Explain your changes.

Creates new directory for soft dependencies namely `torch` and `transformers` in the `util` directory and contains dummy classes for the soft dependencies making them reusable across all vendored libraries.

#### What should a reviewer concentrate their feedback on?

I would like some feedback on the file and `class` structure and if this will be a good style for importing the classes and subclasses in other files.

### CODE REVIEW

**Comment by jgyasu** [comment]
> This looks really good! I believe the current dummy classes are coming from TimeLLM, right? Or is there any other forecaster you considered?

Thanks! And yes, this has just implemented the dummy classes for TimeLLM right now. I have planned to extend this for all the estimators once I'm done working on the TimeLLM.

**Review by fkiraly — CHANGES_REQUESTED** [review]
I think the pattern for `torch` makes sense, since it is a "singular" class that is important.

However, for the classes in `transformers`, this is not what I would do - the pattern is not "extensible" and thus not a very good design - to add a new class to mask, you need to modify a class rather than adding an artefact.

One other option that imo is much simpler is: define a function `_safe_import` or similar, which is used as `foo = _safe_import("torch.foo")`, which internally does "from torch import foo" of the dependency is present - otherwise you return a `Mock` object.

More generally, `bar = _safe_import("torch.foo.foo2")` replaces `from torch.foo import foo2 as bar`. What do you think about this?

To avoid runtime sinks due to failed imports, I would use `_check_soft_dependency` to check for the package before import.

Minor issue: import names are not equal package names, but for this function we may assume that for now, and explicitly mention this in the docstring.

**Comment by fkiraly** [comment]
regarding extensibility, here is an entry point:
https://en.wikipedia.org/wiki/Extensibility
(wikipedia article is ok-ish, but should be a good starting point)

with your current design, imagine what a developer would have to do to vendor a new package depending on `transformers` - they have to vendor the package *plus* add entries to your class. And they would have to know in the first place that they have to do the latter, which seems unlikely unless you are the developer yourself (or unless you write complicated documentation, which is also unlikely to be read).

A better design ensures that the only changes that are needed are in the new code, the vendor package, possibly based off the current module.

**Comment by jgyasu** [comment]
Thanks @fkiraly, this sounds great!

Also I am kind of amazed how senior devs are so far-sighted, I read halfway through your comment and was already imagining the issues that will arise with some package names because of the dot but then I saw your last paragraph :slightly_smiling_face:

> Minor issue: import names are not equal package names, but for this function we may assume that for now, and explicitly mention this in the docstring. Perhaps an optional arg pkg_name, that can be set if it is different - otherwise it is assumed the package name is the string before the first dot.

**Comment by fkiraly** [comment]
> and was already imagining the issues that will arise with some package names because of the dot

but that's exactly what I also did while writing. Top-down design, work through typical cases, edge cases, etc.

**Comment by jgyasu** [comment]
Hi @fkiraly, I have added the `_safe_import` utility in my latest commit. But I have a design concern that I want to discuss.

Currently with how `_safe_import` is designed, suppose if `torch` needs to be imported, then we do something like `torch = _safe_import("torch")` and it uses `_check_soft_dependencies` to check the presence and if it is present then imports it otherwise returns a `Mock` object. This works fine in a situation like this but let's take an another example,
```python
from transformers import (
        BertConfig,
        BertModel,
        BertTokenizer,
        GPT2Config,
        GPT2Model,
        GPT2Tokenizer,
        LlamaConfig,
        LlamaModel,
        LlamaTokenizer,
    )
```
Here,
```python
BertConfig = _safe_import("BertConfig")
BertModel = _safe_import("BertModel")
# ... and so on
```
will return `Mock` objects but if the `transformers` is present in the environment then it won't be able to import these classes like `BertConfig` from `transformers` because it will simply look for `BertConfig` in the environment. I thought of a workaround, what if we pass an additional parameter, let's say we call it `import_subset` and when it is set to `True` then one has to necessarily provide the `pkg_name` parameter as well and then we can do `from pkg_name import import_path`. (Please refer to the commit to see what `pkg_name` and `import_paths` are). 

What do you think about this? FYI @PranavBhatP

**Comment by fkiraly** [comment]
Not sure if I get the situation correctly where the `_safe_import` does not work as intended.

For `BertConfig`, I would do `BertConfig = _safe_import("transformers.BertConfig")`, according to the specification this will do:

1. if `transformers` is present, it is equivalent to `from transformers import BertConfig`
2. if `transformers` is not present, it will be equivalent to `BertConfig = MagicMock()`

**Comment by jgyasu** [comment]
> if transformers is present, it is equivalent to from transformers import BertConfig

Hmm, with how I designed `_safe_import`, this won't work. I found a fix, I can use `getattr`

**Comment by fkiraly** [comment]
> > if transformers is present, it is equivalent to from transformers import BertConfig
> 
> Hmm, with how I designed `_safe_import`, this won't work. I found a fix, I can use `getattr`

I see - that is different from the suggested specification though. Citing myself above:

`bar = _safe_import("torch.foo.foo2")` replaces `from torch.foo import foo2 as bar`. What do you think about this?

That is, imo, the "right" specification. And if there is no dot, it just does `import torch`.

**Comment by jgyasu** [comment]
> > > if transformers is present, it is equivalent to from transformers import BertConfig
> > 
> > 
> > Hmm, with how I designed `_safe_import`, this won't work. I found a fix, I can use `getattr`
> 
> I see - that is different from the suggested specification though. Citing myself above:
> 
> > `bar = _safe_import("torch.foo.foo2")` replaces `from torch.foo import foo2 as bar`. What do you think about this?
> 
> That is, imo, the "right" specification that should solve all common and edge cases. And if there is no dot, it just does `import torch`.

My bad, I missed it. I have modified it, now it should work for all the cases.

**Review by fkiraly — APPROVED** [review]
Looks good!

May I request - non-blocking - to add more clarifying explanation in the docstring?

* In the preamble: what this is equivalent to (like the examples in our discussion)
* for `pkg_name`, give examples (`scikit-learn` vs `sklearn`) to disambiguate package name and import name
* for the string, how it can look like, and what happens if there are multiple or no dots

**Review by fkiraly — APPROVED** [review]
Looks good - I would suggest we run a round of "test all" to make absolutely sure, after merge.
